### ISSUE TITLE
systemd hogs CPU with Dbus events generated in by-diskseq

### ISSUE THREAD

**Issue by zaitcev** [systemd/systemd#25931]
### systemd version the issue has been seen with

252

### Used distribution

Fedora 37

### Linux kernel version used

6.0.15-300.fc37.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd-udevd

### Expected behaviour you didn't see

The box in question used to operate normally until Fedora 36. There was no abnormal CPU consumption by systemd.

### Unexpected behaviour you saw

The initial symptom is, systemd consumes more and more CPU.

We suspect that it has something to do with a broken SD card reader, which disconnects and reconnects  to USB periodically. Each time it happens, something like this is shown in the systemd debug log (journalctl):

Jan 04 12:26:28 guren.zaitcev.lan systemd[1]: dev-disk-by\x2ddiskseq-14732.device: Installed new job dev-disk-by\x2ddiskseq-**14732**.device/nop as 111440716
Jan 04 12:26:28 guren.zaitcev.lan systemd[1]: dev-disk-by\x2ddiskseq-14731.device: Installed new job dev-disk-by\x2ddiskseq-**14731**.device/nop as 111440717
................
Jan 04 12:26:33 guren.zaitcev.lan systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2ddiskseq_2d9_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=389936103 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jan 04 12:26:33 guren.zaitcev.lan systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2ddiskseq_2d9_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=389936104 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a

The number of these jobs, and thus Dbus messages corresponds to this:

[root@guren ~]# cat /sys/devices/pci0000:00/0000:00:1a.0/usb1/1-1/1-1.1/1-1.1:1.0/host6/target6:0:0/6:0:0:0/block/sdb/diskseq
15215
[root@guren ~]# ls /dev/disk/by-diskseq/
1  15215  3
[root@guren ~]# 

So, at this time about 15,000 jobs and messages are generated.

### Steps to reproduce the problem

You should probably be able to find an external SD card reader and plug it several times. Observe the log with journalctl. See that more and more "jobs" are generated for each plug cycle, numbered from 1 to N.

Naturally your system isn't going to fall to its knees just from a couple of plug cycles. It takes some cycling for the number of messages to become severe enough.

**Comment by bluca** [comment]
You can add an udev rule that sets `ID_IGNORE_DISKSEQ` for that specific hardware, that should fix your problem. We can't know what hardware works and what is broken, so not sure anything else can be done.

**Comment by zaitcev** [comment]
The broken hardware makes the problem visible, but the actual bug is that something in depths of udev generates the N events every time a device is (re-)connected, where N is the number of reconnects. If, say, you re-connented 100 times, on the 101st time, 101 events are posted. This is very obviously a problem, which we cannot blame on bad hardware.

**Comment by bluca** [comment]
No, that is how diskseq is supposed to work, the kernel sends notifications and we react accordingly. Did ID_IGNORE_DISKSEQ solve your issue or not?

**Comment by poettering** [comment]
whenever the usb sd reader disconnects/reconnects another set of uevents/media change events are generated, for the device (usually 4 per plug/unplug cycle). These then processed and result in device nodes and symlinks to it be created (and subsequently removed again), by kernel+udev. This is also then propagated to other software, such as systemd and udisks and similar. Which will also act on it.

So, yes, if you have a device that floods the system with uevents lots of things happen, but there's little we can do about that, we have to act on this.

diskseq stuff is a red herring. It's just one more symlink we create, we create a bunch more anyway, and always have. it just becomes a bit more visible because the diskseq device names/symlinks are guaranteed to be different each time, while the other ones should usually come up with the same name if they are the same device plugged into the same port with the same media in it.

Anyway if you have a problem with the kernel generating uevents for devices that userspace then needs to process, please talk to the kernel people. But at some point I am sure that faulty hardware will always be a problem: it will cause software to do stuff it better shouldn't do.

anyway, I see nothing actionable here for us. We work with the events we get. If those are to many for you, then find a way to address that at the source that generates these events, not at the last element in the chain that has no other option than process the firehose it gets from the layers underneath.

hence closing.

**Comment by dagrh** [comment]
> whenever the usb sd reader disconnects/reconnects another set of uevents/media change events are generated, for the device (usually 4 per plug/unplug cycle). These then processed and result in device nodes and symlinks to it be created (and subsequently removed again), by kernel+udev. This is also then propagated to other software, such as systemd and udisks and similar. Which will also act on it.

Isn't what @zaitcev is saying though that he's seeing way more than the 4 per plug/unplug - he's saying the number of events he's seeing is increasing each time it happens; so that on the 'n'th unplug he's seeing something proportional to 'n' events.

**Comment by poettering** [comment]
@dagrh that's not how i understood the report. But even if that's how it was meant: the diskseq is a kernel variable to which userspace has no control, hence if it has been bumped all the way to 15K, then this means 15K block devices or media change events on them have been generated. i.e. it's a kernel issue, not a userspace issue either way.

**Comment by zaitcev** [comment]
The total amount of Dbus events generated is 15000*15000/2. The 15000 is amount generated each time, by now. I'll check how exactly the kernel generates these events and how they are delivered, and I'll re-open this if it turns out a fixed number.

**Comment by yuwata** [comment]
Note, this should be a duplicate of #25021.

**Comment by poettering** [comment]
ok, so if the media change events are just the trigger and not the problem itself, maybe lets reopen this

**Comment by DaanDeMeyer** [comment]
@zaitcev Did you manage to figure out if this is systemd or the kernel?

**Comment by zaitcev** [comment]
@DaanDeMeyer Negative, only disabled the offending device in firmware for now.

**Comment by yuwata** [comment]
So, the reporter disabled the device. And, no more response can be expected.
Let's close this at least now. This should be a driver issue.
